python/ray/util/collective/collective_group/nccl_collective_group.py

```diff
@@ -109,7 +111,43 @@ def get_nccl_id(self, timeout_s=180):
         return uid
 
 
+class StreamPool:
+    """
+    The class that represents a stream pool associated with a device
+
+    Args:
+        size: the size of the pool, which should be fixed across all device
+    """
+    def __init__(self, size):
+        #TODO(Fu): add a lock
+        self.size = size
+        self.pool = [None] * size
+        self.index = 0
+        self.mux = Lock()
+
+    def getStreamFromPool(self):
+        """Get the stream from pool. The function locks first and releases before returning
+
+        Returns:
+            stream (cupy.cuda.Stream): the stream on which the kernel should be placed
+        """
+
+        self.mux.acquire()
+        # Create a new stream if it hasn't been initialized yet
+        if self.pool[self.index] == None:
+            # TODO(Fu): double check whether to set the non_blocking variable
+            self.pool[self.index] = cupy.cuda.Stream(null=False)
+        stream = self.pool[self.index]
+        self.index = (self.index + 1) % self.size
```

Comment:
this way of deciding stream assignment seems too simplified. I believe in PyTorch there is a more complex mechanism for stream prioritization?